akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -171,17 +177,20 @@ abstract class GraphStageLogic {
    * The method [[isAvailable()]] can be used to query if the port has an element that can be grabbed or not.
    */
   final protected def grab[T](in: Inlet[T]): T = {
-    require(isAvailable(in), "Cannot get element from already empty input port")
     val connection = conn(in)
-    val elem = interpreter.connectionStates(connection)
-
-    elem match {
-      case CompletedHasElement(realElem) ⇒
-        interpreter.connectionStates(connection) = Completed
-        realElem.asInstanceOf[T]
-      case _ ⇒
-        interpreter.connectionStates(connection) = Empty
-        elem.asInstanceOf[T]
+    // Fast path
+    if ((interpreter.portStates(connection) & (InReady | InFailed)) == InReady &&
+      (interpreter.connectionSlots(connection).asInstanceOf[AnyRef] ne Empty)) {
```

Comment:
what is kept in `connectionSlots` ? should it be `AnyRef` instead of `Any`? otherwise I don't think there is any overhead of using `!=` here